sdk/trace/provider.go

```diff
@@ -164,7 +179,7 @@ func (p *TracerProvider) Tracer(name string, opts ...trace.TracerOption) trace.T
 func (p *TracerProvider) RegisterSpanProcessor(sp SpanProcessor) {
 	p.mu.Lock()
 	defer p.mu.Unlock()
-	if p.isShutdown {
+	if p.isShutdown.Load() {
```

Comment:
We can also add

```golang
if p.isShutdown.Load() {
  return
```

before aquiring the lock (`p.mu.Lock()`)

Then we will have the "book" implementation of double-checked locking 😉

Reply:
We can but that would change how `RegisterSpanProcessor()`/`UnregisterSpanProcessor()`/`Shutdown()` calls can interleave with each other. There is no need to complicate the code, this change wouldn't do anything useful.

Reply:
Rephrasing @Aneurysm9 comment

In that case there would not be a way for a `SpanProcessor`'s Shutdown() to enter protected sections of `tracer.Shutdown()` as the flag indicating the provider is shut down would already be set.

While it would be strange for a Span Processor shutdown to call `RegisterSpanProcessor` or `UnregisterSpanProcessor` I think we should have "bullet-proof" code which would not lead to deadlocks.

Reply:
> If a library user is doing something unreasonable,

I wouldn't like to judge and think it is hard to imagine all possible scenarios. E.g. maybe in some error scenarios the user would like to unregister a span processor (so `UnregisterSpanProcessor` would be called via `otel.Handle`. Then if it may lead to deadlock.

Personally, I find using concurrency patterns consistently simpler (less complicated). Otherwise, I need to think "why the double-checked locking is done only in one place?".

Anyway I removed "need's work" to unblock you, but also not approving so that other reviewers could comment as well.
Even if my comment is not addressed, it is still an improvement 😉

Reply:
I agree with @pellared here.  Everywhere that should be a no-op after shutdown should first check for the shutdown flag and then take a lock on the mutex only if necessary.

Reply:
Ok, no problem. I've made the change, PTAL.